myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md

```diff
@@ -0,0 +1,662 @@
+        trend = pm.Deterministic("trend", alpha + beta * t, dims="obs_id")
+
+        ## Priors for seasonality
+        beta_fourier = pm.Normal(
+            "beta_fourier",
+            mu=priors["beta_fourier"]["mu"],
+            sigma=priors["beta_fourier"]["sigma"],
+            dims="fourier_features",
+        )
+        fourier_terms = pm.MutableData("fourier_terms", ff)
+        seasonality = pm.Deterministic(
+            "seasonality", pm.math.dot(beta_fourier, fourier_terms), dims="obs_id"
+        )
+
+        mu = ar1 + trend + seasonality
+
+        # The Likelihood
+        outcome = pm.Normal("likelihood", mu=mu, sigma=sigma, observed=y)
+        ## Sampling
+        idata_ar = pm.sample_prior_predictive()
+        if full_sample:
+            idata_ar.extend(pm.sample(samples, random_seed=100, target_accept=0.95))
+            idata_ar.extend(pm.sample_posterior_predictive(idata_ar))
+        else:
+            return idata_ar
+
+    n = prediction_steps - ar_data.shape[0]
+    n_order = 10
+    periods = np.arange(prediction_steps) / 7
+
+    fourier_features_new = pd.DataFrame(
+        {
+            f"{func}_order_{order}": getattr(np, func)(2 * np.pi * periods * order)
+            for order in range(1, n_order + 1)
+            for func in ("sin", "cos")
+        }
+    )
+
+    with AR:
+        # update values of predictors pass in nans for number of predicted steps
+        pm.set_data(
+            new_data={
+                "t": list(range(prediction_steps)),
+                "fourier_terms": fourier_features_new.to_numpy().T,
+                "y": np.concatenate([ar_data, np.array([np.nan for i in range(n)])]),
+            },
+            coords={"obs_id": list(range(prediction_steps))},
+        )
+        # use the updated values and predict outcomes and probabilities:
+        idata_preds = pm.sample_posterior_predictive(
+            idata_ar, var_names=["likelihood"], predictions=True, random_seed=100
+        )
+
+    return idata_ar, idata_preds, AR
+```
+
+```{code-cell} ipython3
+priors_0 = {
+    "coef_0": {"mu": 0.2, "sigma": 0.5, "size": 2},
+    "coef_1": {"mu": 0.2, "sigma": 0.03, "size": 2},
+    "alpha": {"mu": -4, "sigma": 0.1},
+    "beta": {"mu": -0.5, "sigma": 0.1},
+    "beta_fourier": {"mu": 0, "sigma": 0.8},
+    "sigma": 8,
+    "init": {"mu": -4, "sigma": 0.1, "size": 1},
+}
+
+
+idata_t_s, preds_t_s, model = make_latent_AR_trend_seasonal_model(y_t_s, fourier_features, priors_0)
+```
+
+```{code-cell} ipython3
+pm.model_to_graphviz(model)
+```
+
+```{code-cell} ipython3
+az.summary(idata_t_s, var_names=["alpha", "beta", "coef_0", "coef_1", "beta_fourier"])
+```
+
+```{code-cell} ipython3
+plot_fits(idata_t_s, preds_t_s)
+```
+
+We can see here how the the model fit again recovers the broad structure and trend of the data, but in addition we have captured the oscillation of the seasonal effect and projected that into the future.
+
```

Comment:
Maybe you could add some closing remarks here?

Reply:
Added a closing paragraph too